content/blog/marginal-performance-improvements.md

```diff
@@ -0,0 +1,76 @@
++++
+date = "2020-07-13"
+publishdate = "2020-07-13"
+title = "The effort behind marginal performance improvements"
+author = "Kuba Podgorski and Alan Bernstein"
+author_twitter = "gsnark"
+author_img = "2"
+image = "/img/blog/marginal-performance-improvements/banner.jpg"
+overlay_color = "blue" # blue, green, or light
+disable_overlay = false
++++
+
+A look at the amount of effort that we sometimes put into pursuing performance.
+
+## Background
+
+Pilosa is built on the [Roaring bitmaps](https://roaringbitmap.org/) concept, a high-performance technique for operations on compressed sets of bits. It starts by representing all data in terms of bitmaps, each defined as one of three "container types":
+
+- Array: a list of integers indicating set bit positions.
+- Bitmap: a direct binary representation of the bitmap.
+- Run: a list of intervals, each indicating a contiguous "run" of set bits.
+
+For each bitmap, its container type is chosen to minimize storage space, which keeps memory use low. Queries on the data are defined in terms of basic logical operations on pairs of bitmaps. Time-optimized operations are implemented for each pair of container types, which precludes the need for converting back and forth between container types (a drain on both memory and time). 
+
+What we gain in query performance, we lose in added code complexity. This complexity is explained in a previous [blog post](/blog/adding-rle-support/), where we detail how four logical operations, each supporting three container types for both operands, require 27 implementation functions.
+
+Now, I'd like to describe another dimension of complexity: in-place operations.
+
+## The Problem
+
+While this proliferation of type-specific implementations does wonders for speed, it can be suboptimal in terms of memory allocation. In an illustrative example, we might execute a union across two existing rows in a pilosa index, to produce a useful result, in a newly allocated container. In the case that one of the operands can be discarded, there is an opportunity to reduce memory usage by re-using that operand's container: an [in-place](https://en.wikipedia.org/wiki/In-place_algorithm) operation. Of course, this is not always possible, and very often it is much simpler to convert one container to a different type, to make use of an existing implementation.
```

Comment:
```suggestion
While this proliferation of type-specific implementations does wonders for speed, it can be suboptimal in terms of memory allocation. In an illustrative example, we might execute a union across the containers of two existing rows in a Pilosa index. In this case, it's not uncommon for the union of two containers to result in yet another, newly allocated container. But, in the case that one of the operands can be discarded, there is an opportunity to reduce memory usage by re-using that operand's container: an [in-place](https://en.wikipedia.org/wiki/In-place_algorithm) operation.
```

I'm not sure if this suggestion is good either, but this paragraph wasn't especially clear to me.